Fix value of this (#27019)
mdn/content: files/en-us/web/javascript/reference/operators/this/index.md

### this in function contexts

The value of the `this` parameter depends on how the function is called, not on how it's defined.

```js
// An object can be passed as the first argument to 'call'
// or 'apply' and 'this' will be bound to it.
const obj = { a: "Custom" };

// Variables declared with var become properties of 'globalThis'.
var a = "Global";

function whatsThis() {
  return this.a; // 'this' depends on how the function is called
}

> whatsThis(); // 'Global'; the 'this' parameter is bound to globalThis in non–strict mode

Reviewer: ```suggestion
whatsThis(); // 'Global'; the 'this' parameter defaults to globalThis in non–strict mode
```

Author: Done.

obj.whatsThis = whatsThis;
obj.whatsThis(); // 'Custom'; the 'this' parameter is set to obj
```

> Using `call()` and `apply()`, you can bind a value to the `this` parameter as if it's an explicit parameter.

Reviewer: I prefer the original wording (pass the value of `this`), unless there's reason otherwise?

Author: Okay, reverted.

const o = { a: 1, b: 3 };

> // The first argument is bound to the implicit 'this' parameter; the remaining
> // arguments are bound to the explicit parameters.

Reviewer: ```suggestion
// The first argument is bound to the implicit 'this' parameter; the remaining
// arguments are bound to the named parameters.
```

Author: Thanks, applied.

add.call(o, 5, 7); // 16

> // The first argument is bound to the implicit 'this' parameter; the second
> // argument is an array whose members are bound to the explicit parameters.

Reviewer: ```suggestion
// The first argument is bound to the implicit 'this' parameter; the second
// argument is an array whose members are bound to the named parameters.
```

Author: Thanks, applied.

};
```

> We can call `getThisGetter` as a method of `obj`, which sets `this` inside the body to `obj`. The returned function is assigned to a variable `fn`. Now, when calling `fn`, the value of `this` returned is still the one set by the call to `getThisGetter`, which is `obj`. If the returned function is not an arrow function, such calls would cause the `this` value to be `globalThis` in non-strict mode and `undefined` in strict mode.

Reviewer: Let's have an overarching statement that `getThisGetter()` is a non-strict function in the intro of this section. I don't feel it's necessary to split hairs since it's not the point of this example.

Author: Even if `getThisGetter()` is a non-strict function (which it is here), the value of `this` in the `fn()` call can still be `undefined` if the global scope is in strict mode. That’s why the distinction is made.

Reviewer: Yes, I meant `getThisGetter()` is _non-strict_ when it's defined. If the global scope is strict then the function will inherit the strictness.

Author: Exactly, so the `fn()` call will return `globalThis` in non-strict mode and return `undefined` in strict mode, if `getter` was defined as a regular function. So it’s fine or you want to remove that sentence completely?

Author: I see. I find that making no assumption and stating the return values in both cases is still readable and consistent with other descriptions in the documentation, but if you think an assumption would simplify the text let’s do it. How would you state it exactly?

Reviewer: See https://github.com/mdn/content/pull/27019/commits/73fb1d6c79b58f4e204b78098a52006b83e52f37. What do you think?

Author: Thank you @Josh-Cena, it looks good to me.

```js
const fn = obj.getThisGetter();
console.log(fn() === obj); // true
```

> But be careful if you unbind the method of `obj` without calling it, because `getThisGetter` is still a method that has a varying `this` value. Calling `fn2()()` in the following example returns `globalThis` in non-strict mode and `undefined` in strict mode, because it follows the `this` from `fn2()`, which is `globalThis` in non-strict mode and `undefined` in strict mode since it's called without being attached to any object.

Reviewer 2: ```suggestion
But be careful if you unbind the method of `obj` without calling it, because `getThisGetter` is still a method that has a varying `this` value. Calling `fn2()()` in the following example returns `globalThis` in non-strict mode and returns `undefined` in strict mode, because it follows the `this` from `fn2()`, which is `globalThis` in non-strict mode and is `undefined` in strict mode since it's called without being attached to any object.
```

Author: Thanks, applied.

  configurable: true,
});

console.log(o.average, o.sum); // 2 6
```

### this in DOM event handlers

> When a function is used as an event handler, its `this` is bound to the DOM element on which the listener is placed (some browsers do not follow this convention for listeners added dynamically with methods other than {{domxref("EventTarget/addEventListener", "addEventListener()")}}).

Reviewer 2: ```suggestion
When a function is used as an event handler, its `this` parameter is bound to the DOM element on which the listener is placed (some browsers do not follow this convention for listeners added dynamically with methods other than {{domxref("EventTarget/addEventListener", "addEventListener()")}}).
```

I believe that keeping the usage consistent help, even if it might be a bit passive-aggressive, to convey that it is a function parameter here.

Author: I agree, thanks!

### this in inline event handlers

When the code is called from an inline [event handler attribute](/en-US/docs/Web/HTML/Attributes#event_handler_attributes), its `this` is bound to the DOM element on which the listener is placed:

```html
<button onclick="alert(this.tagName.toLowerCase());">Show this</button>
```

The above alert shows `button`. Note, however, that only the outer scope has its `this` bound this way:

```html
<button onclick="alert((function () { return this; })());">
  Show inner this
</button>
```

> In this case, the `this` paramter of the inner function is bound to `globalThis` in non-strict mode and `undefined` in strict mode.

Reviewer 2: ```suggestion
In this case, the `this` parameter of the inner function is bound to `globalThis` in non-strict mode, and is bound to `undefined` in strict mode.
```

Author: Thanks, applied.

> In this case, the `this` parameter of the inner function is bound to `globalThis` in non-strict mode, and is bound to `undefined` in strict mode.

Reviewer: In this example, it's _always_ non-strict because `onclick` is synthesized as a non-strict function, and the inner `function () {}` doesn't have `"use strict"`.

Author: Oh yes you’re right, fixed.
